File: Python/cpx400sp.py
# ============================================================================
# Name        : cpx400sp.py
# Author      : Julian Stiefel (jstiefel@ethz.ch)
# Version     : 1.0.0
# Created on  : 25.08.2020
# Copyright   :
# Description : A class to implement some basic device specific SCPI commands
#               for the Aim TTi CPX400SP PSU
# ============================================================================

import sys

sys.path.append(".")
from tcp_socket import TCP_Socket

class CPX400SP:
    def __init__(self, ip, port):
        try:
            self.cpx = TCP_Socket(ip, port)
            print('Device created')
        except:
            print('Failed to create device')

    def __del__(self):
        del self.cpx

    def get_identification(self):
        response = self.cpx.ask('*IDN?')
        return response

    def set_voltage(self, v):
        command = 'V1 ' + str(v)
        self.cpx.send(command)

    def get_voltage(self):
        voltage = self.cpx.ask('V1O?')
        return voltage

    def set_current(self, i):
        command = 'I1 ' + str(i)
        self.cpx.send(command)

    def get_current(self):
        current = self.cpx.ask('I1O?')
        return current
        

File: src/cpx400sp.cc
//============================================================================
// Name        : cpx400sp.cc
// Author      : Julian Stiefel (jstiefel@ethz.ch)
// Version     : 1.0.0
// Created on  : 24.08.2020
// Copyright   :
// Description : LXI communication with Aim TTi CPX400SP PSU
//============================================================================

#include "../include/lxi_device_control/cpx400sp.h"

using namespace std;

int main(int argc, char **argv) {
    char response[65536];
    int device, length, timeout = 1000;

    // Initialize LXI library
    lxi_init();

    // Connect to LXI device
    device = lxi_connect("192.168.1.131", 9221, "inst0", timeout, RAW);

    // Send SCPI command
    lxi_send(device, "*IDN?", strlen("*IDN?"), timeout);

    // Wait for response
    lxi_receive(device, response, sizeof(response), timeout);

    printf("%s\n", response);

    // Disconnect
    lxi_disconnect(device);

    return 0;
}